aws_lambda_builders/workflows/nodejs_npm/actions.py

```diff
@@ -0,0 +1,92 @@
+"""
+Action to resolve NodeJS dependencies using NPM
+"""
+
+from aws_lambda_builders.actions import BaseAction, Purpose, ActionFailedError
+from aws_lambda_builders.workflows.python_pip.utils import OSUtils
+from .npm import SubprocessNpm, NpmError
+
+import logging
+
+LOG = logging.getLogger(__name__)
+
+
+class NodejsNpmPackAction(BaseAction):
+
+    NAME = 'CopySource'
+    DESCRIPTION = "Packaging source using NPM"
+    PURPOSE = Purpose.COPY_SOURCE
+
+    def __init__(self, artifacts_dir, scratch_dir, manifest_path, runtime, osutils=None, subprocess_npm=None):
+        self.artifacts_dir = artifacts_dir
+        self.manifest_path = manifest_path
+        self.scratch_dir = scratch_dir
+        self.runtime = runtime
+
+        self.osutils = osutils
+        if osutils is None:
+            self.osutils = OSUtils()
+
+        self.subprocess_npm = subprocess_npm
+
+        if self.subprocess_npm is None:
+            self.subprocess_npm = SubprocessNpm(self.osutils)
+
+    def execute(self):
+        try:
+            if not self.osutils.file_exists(self.manifest_path):
+                raise ActionFailedError('package.json not found in: %s' % self.manifest_path)
+
+            package_path = "file:%s" % self.osutils.abspath(self.osutils.dirname(self.manifest_path))
+
+            LOG.debug("NODEJS packaging %s to %s", package_path, self.scratch_dir)
+
+            tarfile_name = self.subprocess_npm.main(['pack', '-q', package_path], cwd=self.scratch_dir)
+
+            LOG.debug("NODEJS packed to %s", tarfile_name)
+
+            tarfile_path = self.osutils.joinpath(self.scratch_dir, tarfile_name)
+
+            self.osutils.extract_tarfile(tarfile_path, tarfile_path + '-unpacked')
+
+            self.osutils.copytree(self.osutils.joinpath(tarfile_path + '-unpacked', 'package'), self.artifacts_dir)
```

Comment:
I'll leave this decision to you but copysource also has additional features such as file filtering which we might not want here (I assume we trust NPM to know what files to include/exclude). This again goes back to the idea to share osutils which I would prefer. We need to have some way of untaring anyway, which requires some form of osutils here.

Reply:
I would stick with CopySource action if you want. You can always set the exclude list to empty which will not exclude anything.

I like to use CopySource action because in the future we might implement `rsync` like capabilities to it, which might copy only the files that changed and not everything. This will come in super handy for hot-reloading.